Add GPU Support to LangChain4j via GPULlama3.java Powered by TornadoVM

mikepapadim opened the pull request:
This pull request integrates GPU-accelerated large language model (LLM) inference into LangChain4j by incorporating GPULlama3.java](https://github.com/beehive-lab/GPULlama3.java), a Java-native implementation of Llama3, Mistral, Phi-3, Qwen2.5, Qwen3 and Disteled-DeepSeek models powered by TornadoVM.

Key Features:
* Multi-Model Support: Llama3, Mistral, Phi-3, Qwen2.5, Qwen3, Distilled-DeepSeek.
* GPU Backends: Supports PTX (NVIDIA) and OpenCL devices (Intel/Apple GPUs).
* CPU Fallback: Switch between GPU and pure-Java CPU execution seamlessly, as GPULlama3.java extends the existing Llama3.java implementation.


## General checklist
<!-- Please double-check the following points and mark them like this: [X] -->
- [x] There are no breaking changes
- [x] I have added unit and/or integration tests for my change
- [ ] The tests cover both positive and negative cases
- [ ] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [x] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green
<!-- Before adding documentation and example(s) (below), please wait until the PR is reviewed and approved. -->
- [x] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [x] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)


## Checklist for adding new maven module
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have added my new module in the root `pom.xml` and `langchain4j-bom/pom.xml`


## Checklist for adding new embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreRemovalIT` that extends from `EmbeddingStoreWithRemovalIT`

## Checklist for changing existing embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

ibriq committed:
[WIP] Implement gpu-llama3 integration

mikepapadim committed:
Revert changes for a2a

mikepapadim committed:
Revert changes

mikepapadim committed:
Add README for langchain4j GPULlama3 integration

mikepapadim committed:
Update pom.xml to add description for GPULlama3 integration and remove unused test dependency

mikepapadim committed:
Add documentation for GPULlama3.java integration

mikepapadim committed:
Update language model integrations docs to include GPULlama3.java link

mikepapadim committed:
Add documentation for GPULlama3.java integration in sidebar

mikepapadim committed:
Refactor class name to `GPULlama3ChatModel` and add method to print last run metrics

mikepapadim committed:
Add GPULlama3StreamingChatModel implementation with builder pattern

mikepapadim committed:
Update pom.xml to include dependency for langchain4j-gpu-llama3

mikepapadim committed:
Refactor `GPULlama3StreamingChatModel` to update response handling and remove unused methods

mikepapadim committed:
Add `onGPU` flag to GPULlama3 models and refactor response handling

mikepapadim committed:
Update gpu-llama3 dependency version to 2.0-SNAPSHOT in pom.xml

dliubarskyi committed:
Merge branch 'main' into gpullama3-integration

dliubarskyi commented on docs/docs/integrations/embedding-models/gpullama3-java.md:
```diff
@@ -0,0 +1,38 @@
+<dependency>
+<groupId>dev.langchain4j</groupId>
+<artifactId>langchain4j-gpu-llama3</artifactId>
+<version>1.4.0-SNAPSHOT</version>
```
```suggestion
<version>1.5.0-beta11</version>
```

dliubarskyi commented on docs/docs/integrations/embedding-models/gpullama3-java.md:
```diff
@@ -0,0 +1,38 @@
+<dependency>
+    <groupId>dev.langchain4j</groupId>
+    <artifactId>langchain4j</artifactId>
+    <version>1.4.0</version>
```
```suggestion
    <version>1.5.0</version>
```

dliubarskyi commented on docs/docs/integrations/embedding-models/gpullama3-java.md:
````diff
@@ -0,0 +1,38 @@
+For Gradle project `build.gradle`
+
+```groovy
+implementation 'dev.langchain4j:langchain4j:1.4.0'
````
```suggestion
implementation 'dev.langchain4j:langchain4j:1.5.0'
```

dliubarskyi commented on docs/docs/integrations/embedding-models/gpullama3-java.md:
````diff
@@ -0,0 +1,38 @@
+
+```groovy
+implementation 'dev.langchain4j:langchain4j:1.4.0'
+implementation 'dev.langchain4j:langchain4j-gpu-llama3:1.4.0'
````
```suggestion
implementation 'dev.langchain4j:langchain4j-gpu-llama3:1.5.0-beta11'
```

dliubarskyi commented on docs/docs/integrations/embedding-models/gpullama3-java.md:
This documentation page is currently under "embedding models" section, but as far as I see, only chat models are implemented

mikepapadim replied:
Thanks for the review. Let me move it in the right section.

dliubarskyi commented on docs/docs/integrations/embedding-models/gpullama3-java.md:
````diff
@@ -0,0 +1,38 @@
+```groovy
+implementation 'dev.langchain4j:langchain4j:1.4.0'
+implementation 'dev.langchain4j:langchain4j-gpu-llama3:1.4.0'
+```
````
A minimal example of creating and running mdoel would be great

mikepapadim replied:
https://github.com/mikepapadim/langchain4j/blob/1ddf49f8bfbeee0aedc26bfbbe4df73f85070673/docs/docs/integrations/language-models/gpullama3-java.md

dliubarskyi commented on langchain4j-bom/pom.xml:
```diff
@@ -201,6 +201,13 @@
+            <dependency>
+                <groupId>dev.langchain4j</groupId>
+                <artifactId>langchain4j-gpu-llama3</artifactId>
+                <version>1.4.0-beta10-SNAPSHOT</version>
```
```suggestion
                <version>1.5.0-beta11-SNAPSHOT</version>
```

dliubarskyi commented on langchain4j-gpu-llama3/pom.xml:
```diff
@@ -0,0 +1,154 @@
+    <parent>
+        <groupId>dev.langchain4j</groupId>
+        <artifactId>langchain4j-parent</artifactId>
+        <version>1.4.0-beta10-SNAPSHOT</version>
```
```suggestion
        <version>1.5.0-beta11-SNAPSHOT</version>
```

dliubarskyi commented on langchain4j-gpu-llama3/pom.xml:
```diff
@@ -0,0 +1,154 @@
+    </parent>
+
+    <artifactId>langchain4j-gpu-llama3</artifactId>
+    <version>1.4.0-SNAPSHOT</version>
```
```suggestion
```

dliubarskyi commented on langchain4j-gpu-llama3/pom.xml:
```diff
@@ -0,0 +1,154 @@
+    <artifactId>langchain4j-gpu-llama3</artifactId>
+    <version>1.4.0-SNAPSHOT</version>
+    <name>LangChain4j :: Integration :: GPULlama3.java</name>
+    <description> GPULlama3.java: GPU-enabled LLM Inference Engine for Java based on TornadoVM- Requires Java 21+</description>
```
```suggestion
    <description>GPULlama3.java: GPU-enabled LLM Inference Engine for Java based on TornadoVM- Requires Java 21+</description>
```

dliubarskyi commented on langchain4j-gpu-llama3/pom.xml:
```diff
@@ -0,0 +1,154 @@
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j-core</artifactId>
+            <version>1.4.0-SNAPSHOT</version>
```
```suggestion
            <version>1.5.0-SNAPSHOT</version>
```

dliubarskyi commented on langchain4j-gpu-llama3/pom.xml:
```diff
@@ -0,0 +1,154 @@
+
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j-http-client</artifactId>
```
why is this dependency required?

dliubarskyi commented on langchain4j-gpu-llama3/pom.xml:
```diff
@@ -0,0 +1,154 @@
+
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j-http-client-jdk</artifactId>
```
why is this dependency required?

mikepapadim committed:
Update docs/docs/integrations/embedding-models/gpullama3-java.md

Co-authored-by: Dmytro Liubarskyi <ljubarskij@gmail.com>

mikepapadim committed:
Update docs/docs/integrations/embedding-models/gpullama3-java.md

Co-authored-by: Dmytro Liubarskyi <ljubarskij@gmail.com>

mikepapadim committed:
Update docs/docs/integrations/embedding-models/gpullama3-java.md

Co-authored-by: Dmytro Liubarskyi <ljubarskij@gmail.com>

dliubarskyi commented on langchain4j-gpu-llama3/pom.xml:
```diff
@@ -0,0 +1,154 @@
+        </dependency>
+
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
```
why is this and other Jackson dependencies required?

mikepapadim committed:
Update docs/docs/integrations/embedding-models/gpullama3-java.md

Co-authored-by: Dmytro Liubarskyi <ljubarskij@gmail.com>

dliubarskyi commented on langchain4j-gpu-llama3/pom.xml:
```diff
@@ -0,0 +1,154 @@
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j-core</artifactId>
+            <version>1.4.0-SNAPSHOT</version>
```
```suggestion
            <version>1.5.0-SNAPSHOT</version>
```
please update to the 1.5.0-SNAPSHOT version everywhere in this pom

dliubarskyi commented on langchain4j-gpu-llama3/src/main/java/dev/langchain4j/model/gpullama3/GPULlama3ChatModel.java:
```diff
@@ -0,0 +1,110 @@
+            this.modelPath = modelPath;
+            return this;
+        }
+        public Builder temperature(float temperature) {
```
```suggestion
        public Builder temperature(double temperature) {
```
please use doubles instead of floats (everywhere)

dliubarskyi commented on langchain4j-gpu-llama3/src/main/java/dev/langchain4j/model/gpullama3/GPULlama3ChatModel.java:
```diff
@@ -0,0 +1,110 @@
+            this.temperature = temperature;
+            return this;
+        }
+        public Builder topp(float topp) {
```
```suggestion
        public Builder topP(float topP) {
```

dliubarskyi commented on langchain4j-gpu-llama3/src/main/java/dev/langchain4j/model/gpullama3/GPULlama3ChatModel.java:
```diff
@@ -0,0 +1,110 @@
+
+    public static class Builder {
+        private Path modelPath;
+        private float temperature = 0.7f;
```
Should all these defaults be here? Will it not work without these defaults?

dliubarskyi commented on langchain4j-gpu-llama3/src/main/java/dev/langchain4j/model/gpullama3/GPULlama3ChatModel.java:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    @Override
+    public String chat(String userMessage) {
```
please implement `doChat(ChatRequest)` method instead, you can see `OpenAiChatModel` implementation as a good example

dliubarskyi commented on langchain4j-gpu-llama3/src/main/java/dev/langchain4j/model/gpullama3/GPULlama3StreamingChatModel.java:
```diff
@@ -0,0 +1,121 @@
+import java.io.IOException;
+import java.nio.file.Path;
+
+public class GPULlama3StreamingChatModel implements StreamingChatModel {
```
same comments as in non-streaming model (above)

dliubarskyi commented on langchain4j-gpu-llama3/src/main/java/dev/langchain4j/model/gpullama3/GPULlama3StreamingChatModel.java:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    @Override
+    public void chat(String userMessage, StreamingChatResponseHandler handler) {
```
please implement `chat(ChatRequest chatRequest, StreamingChatResponseHandler handler)` instead

mikepapadim committed:
Add setup instructions for GPULlama3.java in documentation

dliubarskyi commented on langchain4j-gpu-llama3/src/main/java/dev/langchain4j/model/gpullama3/GPULlama3StreamingChatModel.java:
```diff
@@ -0,0 +1,121 @@
+        handler.onCompleteResponse(chatResponse);
+    }
+
+    private Options getOptions(String userMessage) {
```
perhaps this part can be extracted in a separate method and re-used by both sync and streaming models?

mikepapadim committed:
Remove GPULlama3.java integration documentation file

dliubarskyi reviewed (commented):
@mikepapadim thank you!

mikepapadim committed:
Merge branch 'gpullama3-integration' of github.com:mikepapadim/langchain4j into gpullama3-integration

dliubarskyi reviewed (commented):
You have marked "I have added unit and/or integration tests for my change", but I see no tests in this PR. To make sure your implementation is compliant, please implements all tests mentioned this [PR](https://github.com/langchain4j/langchain4j/pull/3665/files)

mikepapadim commented:
> You have marked "I have added unit and/or integration tests for my change", but I see no tests in this PR. To make sure your implementation is compliant, please implements all tests mentioned this [PR](https://github.com/langchain4j/langchain4j/pull/3665/files)

I added integration tests in the separate langchain-examples repository with a separate PR there.

https://github.com/langchain4j/langchain4j-examples/pull/170

dliubarskyi commented:
@mikepapadim that is not really a test, and it is in another repository, that's not how it works

mikepapadim committed:
Update langchain4j-gpu-llama3/pom.xml

Co-authored-by: Dmytro Liubarskyi <ljubarskij@gmail.com>

mikepapadim committed:
Update langchain4j-gpu-llama3/pom.xml

Co-authored-by: Dmytro Liubarskyi <ljubarskij@gmail.com>

mikepapadim committed:
Update `pom.xml` with version bump to 1.5.0-SNAPSHOT and cleanup unused dependencies

mikepapadim committed:
Refactor variable `topp` to `topP` in GPULlama3 models for clarity and consistency

mikepapadim committed:
Refactor GPULlama3 models to use base class for shared properties and methods; introduce GPULlama3Provider for model and sampler management

mikepapadim committed:
Refactor GPULlama3 models to consolidate initialization in the base class, remove GPULlama3Provider, and streamline response handling logic

mikepapadim committed:
Refactor GPULlama3ChatModel to clean up imports, simplify Builder methods, and streamline chat response handling logic

mikepapadim committed:
Refactor GPULlama3BaseModel to enhance response handling, improve prompt extraction, and add metrics printing functionality

mikepapadim committed:
Refactor GPULlama3BaseModel and GPULlama3StreamingChatModel to enhance response handling, validate chat requests, and streamline Builder methods

mikepapadim commented:
@dliubarskyi I just pushed the refactoring as suggested, only pending is adding the tests.
I am trying to figure out how to add them cause they require someone to have configured [TornadoVM](https://github.com/beehive-lab/TornadoVM) on the system in order to be able to compile the models for the GPU.

mikepapadim committed:
Cleanups GPULlama3BaseModel and GPULlama3StreamingChatModel to clean up imports, remove unused methods, and improve code formatting consistency

mikepapadim committed:
Refactor GPULlama3BaseModel to change `onGPU` and `stream` types to `Boolean` and update model loading to use `onGPU` parameter

mikepapadim committed:
Apply suggestion from @dliubarskyi

Co-authored-by: Dmytro Liubarskyi <ljubarskij@gmail.com>

mikepapadim committed:
[WIP] Add integration tests for GPULlama3ChatModel and GPULlama3StreamingChatModel with GPU support disabled by default

mikepapadim committed:
Update GPULlama3.java documentation with expanded setup, model compatibility, and usage examples

mikepapadim committed:
Merge branch 'gpullama3-integration' of github.com:mikepapadim/langchain4j into gpullama3-integration

mikepapadim commented:
[docs and instructons](https://github.com/mikepapadim/langchain4j/blob/1ddf49f8bfbeee0aedc26bfbbe4df73f85070673/docs/docs/integrations/language-models/gpullama3-java.md) also updated

dliubarskyi commented:
@mikepapadim we could add a separate GitHub workflow that would use a TornadoVM, would that work?

mikepapadim commented:
> @mikepapadim we could add a separate GitHub workflow that would use a TornadoVM, would that work?

thank you for the reply.
That would work for sure.

In the documentation, I added all the steps 

> @mikepapadim we could add a separate GitHub workflow that would use a TornadoVM, would that work?

Thank you for the reply, that would work for sure.

In the documentation, both here and in the example repo, I added detailed instructions how I managed to run the following examples: https://github.com/mikepapadim/langchain4j-examples/blob/c694832c459c3534410e7322b5f85eee30471de6/gpullama3.java-example/src/main/java/GPULlama3ChatModelExamples.java

How do you want to proceed with this one?

dliubarskyi commented:
@mikepapadim feel free to create a new workflow, similar to `.github/workflows/nightly_jdk21.yaml`, that will run every night with TornadoVM and run ITs for your module

mikepapadim committed:
Update integration tests for GPULlama3ChatModel and GPULlama3StreamingChatModel: enable GPU tagging, update model paths, and rename test class with IT suffix

mikepapadim committed:
Update `pom.xml` for GPULlama3: add TornadoVM dependencies, test runner configurations, and new Maven profiles and plugins

mikepapadim committed:
Expand GPULlama3.java documentation: add project setup instructions, model compatibility details, chat usage examples, testing guidelines, and runtime configuration steps.

mikepapadim committed:
Exclude `langchain4j-gpu-llama3` module from nightly JDK17 and JDK21 workflows as it required special config as ollama

mikepapadim committed:
Spotless apply on GPULlama3 module: clean up imports, enhance code formatting, reorganize `pom.xml`, and improve test structure consistency.

mikepapadim committed:
Remove unnecessary blank line in `langchain4j-bom/pom.xml`.

mikepapadim committed:
Update `gpu-llama3` dependency version to `2.1-SNAPSHOT` in `langchain4j-gpu-llama3/pom.xml`.

mikepapadim commented:
> @mikepapadim feel free to create a new workflow, similar to `.github/workflows/nightly_jdk21.yaml`, that will run every night with TornadoVM and run ITs for your module

@dliubarskyi  Thank you for the suggestion about creating a nightly workflow similar to `nightly_jdk21.yaml`.
I've included[ unit tests](https://github.com/mikepapadim/langchain4j/blob/afda6b0f3ec2455c5fa5066bcff68f908158481f/langchain4j-gpu-llama3/src/test/java/GPULlama3ChatModelIT.java) with [instructions](https://github.com/mikepapadim/langchain4j/blob/afda6b0f3ec2455c5fa5066bcff68f908158481f/langchain4j-gpu-llama3/README.md) for this integration, which I coverage ChatModel and StreamChatModels. However, I'm hesitant to add a GPU-enabled nightly workflow as part of this PR.

Looking at the existing workflows in the repository, I don't see any that currently use GPU resources. They all run on standard GitHub-hosted runners. Adding GPU runners could introduce resource monitoring challenges and cost implications. I also noticed that the ollama module is excluded from the nightly workflow I assume for a similar reason, which suggests GPU-dependent tests  seen to be excluded in CI at the moment.

I'm not sure how to proceed with GPU resource requirements, as this would require an iteration in the curent CI/CD infrastructure. 

Would it make sense to move forward with the unit tests for now and revisit GPU-enabled workflows separately if the it's necessary?

LizeRaes commented:
@mikepapadim first of all, thank you so much for this PR! This is a big step forward for running local models efficiently!
@dliubarskyi regarding GPU runners, I could reach out to GH and see if they could make some resources available given the size of our OSS project? (not saying you have to wait with merging!)

mikepapadim commented:
> @mikepapadim first of all, thank you so much for this PR! This is a big step forward for running local models efficiently! @dliubarskyi regarding GPU runners, I could reach out to GH and see if they could make some resources available given the size of our OSS project? (not saying you have to wait with merging!)

Cool thanks @LizeRaes, we are preparing some cool demos with it for the Devoxx in Belgium

dliubarskyi commented:
@LizeRaes sounds great, thank you!

I will try to review and merge this ASAP without waiting for GPU runners

mikepapadim committed:
[sync] Update `gpu-llama3` dependency to `0.2.1-SNAPSHOT`, fix sampler method call, and correct README Maven profile usage

orionpapadakis committed:
Add slf4j dependencies to pom

dliubarskyi committed:
Merge branch 'main' into gpullama3-integration

dliubarskyi commented on .github/workflows/nightly_jdk17.yaml:
```diff
@@ -15,7 +15,7 @@ jobs:
         include:
           - java_version: '17'
-            mvn_opts: '-pl !langchain4j-ollama,!langchain4j-local-ai,!langchain4j-milvus'
+            mvn_opts: '-pl !langchain4j-ollama,!langchain4j-local-ai,!langchain4j-milvus, !langchain4j-gpu-llama3'
```
```suggestion
            mvn_opts: '-pl !langchain4j-ollama,!langchain4j-local-ai,!langchain4j-milvus,!langchain4j-gpu-llama3'
```

dliubarskyi commented on .github/workflows/nightly_jdk21.yaml:
```diff
@@ -15,7 +15,7 @@ jobs:
         include:
           - java_version: '21'
-            mvn_opts: '-pl !langchain4j-ollama,!langchain4j-local-ai,!langchain4j-milvus'
+            mvn_opts: '-pl !langchain4j-ollama,!langchain4j-local-ai,!langchain4j-milvus, !langchain4j-gpu-llama3'
```
```suggestion
            mvn_opts: '-pl !langchain4j-ollama,!langchain4j-local-ai,!langchain4j-milvus,!langchain4j-gpu-llama3'
```

dliubarskyi commented on docs/docs/integrations/language-models/gpullama3-java.md:
```diff
@@ -0,0 +1,228 @@
+<dependency>
+    <groupId>dev.langchain4j</groupId>
+    <artifactId>langchain4j</artifactId>
+    <version>1.5.0</version>
```
```suggestion
    <version>1.6.0</version>
```

dliubarskyi commented on docs/docs/integrations/language-models/gpullama3-java.md:
```diff
@@ -0,0 +1,228 @@
+<dependency>
+<groupId>dev.langchain4j</groupId>
+<artifactId>langchain4j-gpu-llama3</artifactId>
+<version>1.5.0-beta11</version>
```
```suggestion
<version>1.6.0-beta12</version>
```

dliubarskyi commented on docs/docs/integrations/language-models/gpullama3-java.md:
````diff
@@ -0,0 +1,228 @@
+For Gradle project `build.gradle`
+
+```groovy
+implementation 'dev.langchain4j:langchain4j:1.5.0'
````
```suggestion
implementation 'dev.langchain4j:langchain4j:1.6.0'
```

dliubarskyi commented on docs/docs/integrations/language-models/gpullama3-java.md:
````diff
@@ -0,0 +1,228 @@
+
+```groovy
+implementation 'dev.langchain4j:langchain4j:1.5.0'
+implementation 'dev.langchain4j:langchain4j-gpu-llama3:1.5.0-beta11'
````
```suggestion
implementation 'dev.langchain4j:langchain4j-gpu-llama3:1.6.0-beta12'
```

dliubarskyi commented on docs/docs/integrations/language-models/gpullama3-java.md:
```diff
@@ -0,0 +1,228 @@
+        System.out.println("Example Prompt: " + prompt);
+    }
+
+    // @formatter:off
```
```suggestion
```

dliubarskyi commented on docs/docs/integrations/language-models/gpullama3-java.md:
```diff
@@ -0,0 +1,228 @@
+            .onGPU(Boolean.TRUE) // if false, runs on CPU though a lightweight implementation of llama3.java
+            .modelPath(modelPath)
+            .build();
+    // @formatter:on
```
```suggestion
```

dliubarskyi commented on langchain4j-bom/pom.xml:
```diff
@@ -201,6 +201,12 @@
+            <dependency>
+                <groupId>dev.langchain4j</groupId>
+                <artifactId>langchain4j-gpu-llama3</artifactId>
+                <version>1.5.0-beta11-SNAPSHOT</version>
```
```suggestion
                <version>1.6.0-beta12-SNAPSHOT</version>
```

dliubarskyi commented on langchain4j-gpu-llama3/README.md:
This seems to be a duplicate, please remove

dliubarskyi commented on langchain4j-gpu-llama3/pom.xml:
```diff
@@ -0,0 +1,225 @@
+    <parent>
+        <groupId>dev.langchain4j</groupId>
+        <artifactId>langchain4j-parent</artifactId>
+        <version>1.5.0-beta11-SNAPSHOT</version>
```
```suggestion
        <version>1.6.0-beta12-SNAPSHOT</version>
```

dliubarskyi commented on langchain4j-gpu-llama3/pom.xml:
```diff
@@ -0,0 +1,225 @@
+    </parent>
+
+    <artifactId>langchain4j-gpu-llama3</artifactId>
+    <version>1.5.0-SNAPSHOT</version>
```
```suggestion
```

dliubarskyi committed:
Apply suggestions from code review

dliubarskyi commented on langchain4j-gpu-llama3/README.md:
```diff
@@ -0,0 +1,280 @@
+<dependency>
+    <groupId>dev.langchain4j</groupId>
+    <artifactId>langchain4j</artifactId>
+    <version>1.5.0</version>
```
```suggestion
    <version>1.6.0</version>
```

dliubarskyi commented on langchain4j-gpu-llama3/README.md:
```diff
@@ -0,0 +1,280 @@
+<dependency>
+<groupId>dev.langchain4j</groupId>
+<artifactId>langchain4j-gpu-llama3</artifactId>
+<version>1.5.0-beta11</version>
```
```suggestion
<version>1.6.0-beta12</version>
```

dliubarskyi commented on langchain4j-gpu-llama3/README.md:
````diff
@@ -0,0 +1,280 @@
+For Gradle project `build.gradle`
+
+```groovy
+implementation 'dev.langchain4j:langchain4j:1.5.0'
````
```suggestion
implementation 'dev.langchain4j:langchain4j:1.6.0'
```

dliubarskyi commented on langchain4j-gpu-llama3/README.md:
````diff
@@ -0,0 +1,280 @@
+
+```groovy
+implementation 'dev.langchain4j:langchain4j:1.5.0'
+implementation 'dev.langchain4j:langchain4j-gpu-llama3:1.5.0-beta11'
````
```suggestion
implementation 'dev.langchain4j:langchain4j-gpu-llama3:1.6.0-beta12'
```

dliubarskyi commented on langchain4j-gpu-llama3/pom.xml:
```diff
@@ -0,0 +1,224 @@
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j-core</artifactId>
+            <version>1.5.0-SNAPSHOT</version>
```
```suggestion
            <version>1.6.0-SNAPSHOT</version>
```

dliubarskyi commented on langchain4j-gpu-llama3/pom.xml:
```diff
@@ -0,0 +1,224 @@
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j-core</artifactId>
+            <version>1.5.0-SNAPSHOT</version>
```
```suggestion
            <version>1.6.0-SNAPSHOT</version>
```

dliubarskyi commented on langchain4j-gpu-llama3/pom.xml:
```diff
@@ -0,0 +1,224 @@
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j</artifactId>
+            <version>1.5.0-SNAPSHOT</version>
```
```suggestion
            <version>1.6.0-SNAPSHOT</version>
```

dliubarskyi commented on langchain4j-gpu-llama3/pom.xml:
```diff
@@ -0,0 +1,224 @@
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j</artifactId>
+            <version>1.5.0-SNAPSHOT</version>
```
```suggestion
            <version>1.6.0-SNAPSHOT</version>
```

dliubarskyi commented on langchain4j-gpu-llama3/pom.xml:
```diff
@@ -0,0 +1,224 @@
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j-open-ai</artifactId>
+            <version>1.5.0-SNAPSHOT</version>
```
```suggestion
            <version>1.6.0-SNAPSHOT</version>
```

dliubarskyi committed:
Apply suggestions from code review

dliubarskyi commented on langchain4j-gpu-llama3/pom.xml:
```diff
@@ -0,0 +1,224 @@
+        <dependency>
+            <groupId>org.beehive.gpullama3</groupId>
+            <artifactId>gpu-llama3</artifactId>
+            <version>0.2.1-SNAPSHOT</version>
```
I am not able to find this artifact on the maven central. Where do you publish?

mikepapadim replied:
https://central.sonatype.com/artifact/io.github.beehive-lab/gpu-llama3/overview

dliubarskyi replied:
Nice, thanks for publishing it!

dliubarskyi commented on langchain4j-gpu-llama3/src/test/java/GPULlama3CStreamingChatModelIT.java:
```diff
@@ -0,0 +1,77 @@
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.api.Test;
+
+public class GPULlama3CStreamingChatModelIT {
```
These tests are still not following the approach I mentioned in [this comment](https://github.com/langchain4j/langchain4j/pull/3654#pullrequestreview-3205499786).

We have an extensive set of compliance tests which every LLM integration implements and passes. Please make sure you implement it as well, you only need to extend a few classes and override a few methods, there are plenty of examples in other integrations.

mikepapadim replied:
I followed how Jlama module implements their test. The thing is that many of the common tests under `AbstractBaseChatModelIT` are not supported for GPU inference atm.

dliubarskyi replied:
> I followed how Jlama module implements their test.

Jlama is not the best example when it comes to testing.

> The thing is that many of the common tests under AbstractBaseChatModelIT are not supported for GPU inference atm.

What exactly is not supported? `AbstractBaseChatModelIT` has lots of `supportsXyz()` methods which you can override if something is not supported.

`AbstractBaseChatModelIT` provides a lot of tests out of the box. Current testing is not sufficient. I see some bugs in your code, and those bugs wouldn’t be there if you had followed the guidelines.

mikepapadim committed:
Update `gpu-llama3` dependency groupId and version in `langchain4j-gpu-llama3/pom.xml`

mikepapadim committed:
Update README instructions for running integration tests with TornadoVM GPU acceleration

mikepapadim committed:
Refactor `GPULlama3ChatModelIT` and `GPULlama3CStreamingChatModelIT` to extend shared abstractions, streamline model creation, and enhance test consistency.

mikepapadim committed:
Merge branch 'main' of github.com:langchain4j/langchain4j into gpullama3-integration

mikepapadim committed:
Remove `AbstractChatModelIT` and `AbstractStreamingChatModelIT` extensions from GPU LLaMA3 chat integration tests; cleanup imports and unused overrides.

mikepapadim committed:
Merge branch 'gpullama3-integration' of github.com:mikepapadim/langchain4j into gpullama3-integration

dliubarskyi commented on langchain4j-gpu-llama3/src/main/java/dev/langchain4j/model/gpullama3/GPULlama3ChatModel.java:
```diff
@@ -0,0 +1,103 @@
+    @Override
+    public ChatResponse doChat(ChatRequest chatRequest) {
+        try {
+            // Create and return chat response
```
Validation of `ChatRequest` is missing (like in `GPULlama3StreamingChatModel.doChat()`)

mikepapadim committed:
Add validation for chat request messages and parameters in `GPULlama3ChatModel`.

mikepapadim committed:
[test] First iteration on extending AbstractBaseChatModelIT for GPULlama3ChatModelIT and GPULlama3CStreamingChatModelIT

orionpapadakis committed:
Merge branch 'gpullama3-integration' of github.com:mikepapadim/langchain4j into demo-integration

mikepapadim committed:
Disable parallel execution for JUnit tests in GPULlama3 module.

orionpapadakis committed:
Merge branch 'gpullama3-integration' of github.com:mikepapadim/langchain4j into demo-integration

orionpapadakis committed:
Fixes in model init and free in GPULlama3ChatModelIT and GPULlama3CStreamingChatModelIT

orionpapadakis committed:
Handle the system prompt absence as empty string instead of null

orionpapadakis committed:
Improve flexibility by moving stuff to init (i.e. weight loading)

orionpapadakis committed:
Drop Options usage

orionpapadakis committed:
Refactor GPULlama3 models to remove `Options` dependency and streamline `init` method parameters.

orionpapadakis committed:
Showcase temperature, topP, maxTokens, and seed parameters usage in GPULlama3 model tests

orionpapadakis committed:
Remove unused fields and methods from GPULlama3 model builders

orionpapadakis committed:
Refactor GPULlama3 models to improve response handling, logging, and token processing.

orionpapadakis committed:
Improve GPULlama3ChatModel response parsing with support for `<think>` tags.

orionpapadakis committed:
Refactor response parsing by extracting `GPULlama3ResponseParser` and updating `GPULlama3ChatModel` to use it.

orionpapadakis committed:
Add comprehensive documentation to GPULlama3 classes (`ResponseParser`, `BaseModel`, `ChatModel`) and prevent `ResponseParser` instantiation.

orionpapadakis committed:
Enhance `GPULlama3ResponseParser` to support streaming responses with real-time thinking content separation, preserve `<think>` tags, and add a `StreamingParser` utility. Update related models to integrate streaming functionality and improve error handling.

orionpapadakis committed:
Add `<p>` tags to JavaDocs in GPULlama3 classes for improved formatting and consistency.

orionpapadakis committed:
Enable automatic and manual GPU resource cleanup in `GPULlama3BaseModel` using Java Cleaner API. Implement `AutoCloseable` for try-with-resources support and add resource cleanup safety checks.

dliubarskyi committed:
Merge branch 'refs/heads/main' into fork/mikepapadim/gpullama3-integration

# Conflicts:
#	.github/workflows/nightly_jdk17_1.yaml
#	.github/workflows/nightly_jdk21.yaml

dliubarskyi committed:
versions

dliubarskyi committed:
workflows

dliubarskyi committed:
versions

dliubarskyi committed:
workflows

dliubarskyi committed:
build only with java 21+

dliubarskyi committed:
build only with java 21+

dliubarskyi commented on langchain4j-gpu-llama3/pom.xml:
```diff
@@ -0,0 +1,223 @@
+        <skipTests>true</skipTests>
+        <skipITs>true</skipITs>
+        <maven.compiler.source>21</maven.compiler.source>
+        <maven.compiler.target>21</maven.compiler.target>
```
```suggestion
```
why do you need to set `target` if you already set `release`?

dliubarskyi commented on langchain4j-gpu-llama3/pom.xml:
```diff
@@ -0,0 +1,223 @@
+
+    <properties>
+        <tornado.sdk>${env.TORNADO_SDK}</tornado.sdk>
+        <test.class>GPULlama3ChatModelIT</test.class>
```
```suggestion
```
this property does not seem ot be used anywhere

mikepapadim replied:
this is indeed no longer neeed

dliubarskyi commented on langchain4j-gpu-llama3/pom.xml:
```diff
@@ -0,0 +1,223 @@
+        <maven.compiler.testSkip>true</maven.compiler.testSkip>
+        <skipTests>true</skipTests>
+        <skipITs>true</skipITs>
+        <maven.compiler.source>21</maven.compiler.source>
```
```suggestion
```
why do you need to set `source` if you already set `release`?

dliubarskyi commented on langchain4j-gpu-llama3/pom.xml:
```diff
@@ -0,0 +1,223 @@
+
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j-open-ai</artifactId>
```
this dependency should not be required

mikepapadim replied:
this is not required, I ll rm it

dliubarskyi commented on langchain4j-gpu-llama3/pom.xml:
```diff
@@ -0,0 +1,223 @@
+
+        <dependency>
+            <groupId>com.squareup.okhttp3</groupId>
+            <artifactId>mockwebserver</artifactId>
```
why is this required?

mikepapadim replied:
this is not required, I ll rm it

dliubarskyi commented on langchain4j-gpu-llama3/pom.xml:
```diff
@@ -0,0 +1,223 @@
+
+        <dependency>
+            <groupId>org.junit.platform</groupId>
+            <artifactId>junit-platform-console-standalone</artifactId>
```
why is this required?

mikepapadim replied:
required to run the `ConsoleLauncher` in order to run the module scoped `run-tests` profile to test the IT standolone tests provided

dliubarskyi commented on langchain4j-gpu-llama3/src/test/java/GPULlama3CStreamingChatModelIT.java:
```diff
@@ -0,0 +1,292 @@
+    }
+
+    @AfterEach
+    void tearDown() {
```
this method does not make any sense for me, could you please explain why do you need to nullify the reference and why do you call gc? Why is `model` static?

mikepapadim replied:
this is removed now and part of the logic of `GPULlama3BaseModel` without any explicit `gc` calls. TornadoVM takes care to free the resources on the GPU.

dliubarskyi commented on langchain4j-gpu-llama3/src/main/java/dev/langchain4j/model/gpullama3/GPULlama3BaseModel.java:
```diff
@@ -0,0 +1,123 @@
+        return request.messages().stream()
+                .filter(m -> m instanceof dev.langchain4j.data.message.UserMessage)
+                .map(m -> ((dev.langchain4j.data.message.UserMessage) m).singleText())
+                .reduce((first, second) -> second) // take the last user message
```
Looking more closely into the implementation, it does not seem like `ChatModel` and `StreamingChatModel` are the right  APIs for GPULlama. If all previous messages are **_silently ignored_**, then it is definitely not a _**chat**_ model. Perhaps `LanguageModel` and `StreamingLanguageModel` are better fits?

mikepapadim replied:
We have implemented this as `ChatModel` and `StreamingChatModel`  because in the documentation available it was stated the following:

[Support for LanguageModels will no longer be expanded in LangChain4j, so in all new features, we will use a ChatModel API.](https://docs.langchain4j.dev/tutorials/chat-and-language-models#:~:text=Support%20for%20LanguageModels%20will%20no%20longer%20be%20expanded%20in%20LangChain4j%2C%20so%20in%20all%20new%20features%2C%20we%20will%20use%20a%20ChatModel%20API.)

dliubarskyi replied:
That statement means that all other LC4j APIs (e.g., AI Services) will be built against `ChatModel` API instead of `LanguageModel`, but it does not mean that there will be no `LanguageModel` implementations. I should probably clarify it better in the docu.

Do you have a plan to expand GPULlama3 API to accept a list of messages? This is the API that all modern LLM providers use.

mikepapadim replied:
Yes, we plan to extend it to support list of messages as well. 

I have another question. Currently, in the implementation we support stateless model, should we keep the implementation this way or make it statefull instead?

In there a general guidance? For example, all models providers to be stateless and provide statefull functionality through `Memory` or `AIServices`?

dliubarskyi replied:
Currently all `ChatModel`s are stateless and will remain so, there are not many providers that allow storing state on their side, and the shape of the API is quite different. State is currently stored via `ChatMemory`, yes

mikepapadim replied:
Our `ChatModel` n `StreamingChatModel` implementation are  stateless and will implement statefull behaviour through Memory and AIServices.

mikepapadim committed:
Update langchain4j-gpu-llama3/pom.xml

Co-authored-by: Dmytro Liubarskyi <ljubarskij@gmail.com>

mikepapadim committed:
Update langchain4j-gpu-llama3/pom.xml

Co-authored-by: Dmytro Liubarskyi <ljubarskij@gmail.com>

mikepapadim committed:
Update langchain4j-gpu-llama3/pom.xml

Co-authored-by: Dmytro Liubarskyi <ljubarskij@gmail.com>

dliubarskyi committed:
Merge branch 'refs/heads/main' into fork/mikepapadim/gpullama3-integration

# Conflicts:
#	.github/workflows/nightly_jdk17_1.yaml
#	.github/workflows/nightly_jdk21.yaml
#	.github/workflows/nightly_jdk25.yaml

dliubarskyi committed:
removed unnecessary deps

dliubarskyi committed:
versions

dliubarskyi committed:
docu

dliubarskyi committed:
typo

dliubarskyi committed:
docu

dliubarskyi committed:
slight refactoring

dliubarskyi commented:
@mikepapadim I am planning to do a release tomorrow, around noon. While I am not happy with the current state (when it comes to testing), I would still include it in the release, so that you can demo it on Devoxx without friction.
Since I cannot test it on my side (no GPU) and we still cannot test it on CI, please test it once again on your side (as of current state, I did a few commits today) and confirm that it works as expected. Afterwards I can merge it and include in the release tomorrow. Thanks!

mikepapadim committed:
Refactor to make GPULlama3 models stateless, rename variables for clarity, and improve code consistency.

mikepapadim committed:
Merge branch 'gpullama3-integration' of github.com:mikepapadim/langchain4j into demo-integration

mikepapadim committed:
Remove redundant `@AfterAll` methods and cleanup unused imports in GPULlama3 chat model tests.

mikepapadim committed:
Refactor `GPULlama3BaseModel` to streamline prompt handling logic with `processPromptMessages`, clean up imports, and enhance code clarity in tests.

mikepapadim committed:
Reformat code in `GPULlama3BaseModel` to improve readability by adjusting line breaks and reorganizing imports.

orionpapadakis committed:
Merge pull request #1 from mikepapadim/demo-integration

Demo integration

mikepapadim commented:
> @mikepapadim I am planning to do a release tomorrow, around noon. While I am not happy with the current state (when it comes to testing), I would still include it in the release, so that you can demo it on Devoxx without friction. Since I cannot test it on my side (no GPU) and we still cannot test it on CI, please test it once again on your side (as of current state, I did a few commits today) and confirm that it works as expected. Afterwards I can merge it and include in the release tomorrow. Thanks!

@dliubarskyi  Thank you very much for all the effort putting to merge this!

We have prepared multiple demos for Devoxx. I just tested your latest commits and pushed our latest changes.

Everything works as expected in our end.

The IT tests worked fine in my configuration (adding a screenshot) with an RTX3070.
Also, we tested a few of the [demos we planned to show] (https://github.com/mikepapadim/devoxx25-demo-gpullama3-langchain4j) and worked fine. 
<img width="1292" height="737" alt="image" src="https://github.com/user-attachments/assets/77953964-f7f4-4c29-8022-b88d52919900" />

Also, we [provided a fix](https://github.com/langchain4j/langchain4j/pull/3654/files#diff-a1d689884ce01f7673568973a1eeb545a864959ba38f12c15aeafe8ca311b4e9R187) with full support of multiple messages of System,User, Assistant in `ChatModel` and `StreamingChatModel` .

After Devoxx we will iterate over to implement the missing test and streamlist the testing process. 

Feel free to include it in the release, I can confirm it works as expected!

mikepapadim committed:
Add newline for readability in `langchain4j-bom/pom.xml` dependencies section

mikepapadim committed:
Add newline for readability in `langchain4j-bom/pom.xml` dependencies section

mikepapadim committed:
Bump `gpu-llama3` dependency version to 0.2.2 in `langchain4j-gpu-llama3/pom.xml`.

dliubarskyi commented on langchain4j-gpu-llama3/pom.xml:
```diff
@@ -35,6 +35,19 @@
+        <!-- logging dependencies -->
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-api</artifactId>
```
I do not see any use of slf4j in your module, why is this dependency required?

dliubarskyi commented on langchain4j-gpu-llama3/pom.xml:
```diff
@@ -35,6 +35,19 @@
+
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-simple</artifactId>
```
It is a bad practice to include logging backend in the library, let user do that

dliubarskyi reviewed (approved):
@mikepapadim thanks a lot!

mikepapadim commented:
> @mikepapadim thanks a lot!

Thanks again for the detailed reviews!

We will improve the integration n testing once we are done with Devoxx
